### PR TITLE
[isoltest] test-hooks: add support for "reactions".

### PR DESCRIPTION
**Pull request by aarlt**

### CODE REVIEW

File: test/libsolidity/util/TestFileParser.cpp [hunk 1]
```diff
@@ -674,6 +694,19 @@ string TestFileParser::Scanner::scanHexNumber()
 	return number;
 }
 
+string TestFileParser::Scanner::scanReactionString()
+{
+	string str;
+	advance();
+
+	while (current() != '.')
```
Comment (aarlt) [inline test/libsolidity/util/TestFileParser.cpp:702]: Not sure about this. I chose this "terminator" because the reactions would then look like sentences.
```
// smoke.action ->
// - Hello this is a reaction.
// - Event "A(address,bytes32,uint256)" contained 0x1212121212121212121212121212120000000012, 0x1234, 0x17.
```
Of course this also means that we cannot use a '.' within a reaction.

File: test/libsolidity/SemanticTest.cpp [hunk 2]
```diff
@@ -61,6 +61,7 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 	addBuiltin("smoke.test1", simpleSmokeBuiltin);
 	addBuiltin("smoke.test2", simpleSmokeBuiltin);
 	addBuiltin("smoke_test3", simpleSmokeBuiltin);
+	addBuiltin("smoke.reaction", bind(&SemanticTest::builtinSmokeReactionTest, this, _1));
```
Comment (aarlt) [inline test/libsolidity/SemanticTest.cpp:64]: Hmm.. this test-builtin should be named `smoke.action`.

File: test/libsolidity/util/TestFileParser.cpp (hunk 1 above)
Comment (aarlt) [reply test/libsolidity/util/TestFileParser.cpp:702]: maybe ';' is better

**Comment by chriseth** [comment]
Maybe "effect' is a better term than "reaction"?
Also, does this really depend on the test hooks or could you rebase it in top of develop?

File: test/libsolidity/SemanticTest.cpp [hunk 3]
```diff
@@ -300,7 +301,11 @@ TestCase::TestResult SemanticTest::runTest(
 		}
 
 		for (std::unique_ptr<TestHook> const& hook: m_testHooks)
-			hook->afterFunctionCall(test);
+		{
+			vector<string> reactions = hook->afterFunctionCall(test);
```
Comment (chriseth) [inline test/libsolidity/SemanticTest.cpp:305]: Can't we use a specific function just for the reactions/effects?

File: test/libsolidity/SemanticTest.cpp [hunk 4]
```diff
@@ -452,3 +457,16 @@ std::optional<bytes> SemanticTest::builtinSmokeTest(FunctionCall const& call)
 	}
 	return result;
 }
+
+std::optional<bytes> SemanticTest::builtinSmokeReactionTest(FunctionCall const& call)
```
Comment (chriseth) [inline test/libsolidity/SemanticTest.cpp:461]: Why is this needed/

File: test/libsolidity/util/TestFileParser.cpp [hunk 5]
```diff
@@ -271,6 +273,19 @@ FunctionCallArgs TestFileParser::parseFunctionCallArguments()
 	return arguments;
 }
 
+vector<string> TestFileParser::parseFunctionCallReactions()
+{
+	vector<string> result;
+	while (accept(Token::Reaction, false))
```
Comment (chriseth) [inline test/libsolidity/util/TestFileParser.cpp:279]: Shouldn't this parse `-` instead of the `reaction` keyword?

File: test/libsolidity/util/TestFileParser.cpp (hunk 1 above)
Comment (chriseth) [reply test/libsolidity/util/TestFileParser.cpp:702]: Wouldn't newline be the natural terminator?

File: test/libsolidity/util/TestFunctionCall.cpp [hunk 6]
```diff
@@ -225,6 +225,17 @@ string TestFunctionCall::format(
...
 			for (auto& hook: *_hooks)
 				stream << hook->formatFunctionCall(*this, _errorReporter, _linePrefix, _renderResult, _highlight);
+
+		if (!m_reactions.empty())
+			stream << std::endl;
+
+		for (string const& reaction: m_reactions)
+		{
+			soltestAssert(reaction.find('.') == string::npos, "'.' not allowed in reaction.");
+			stream << _linePrefix << "// - " << reaction << ".";
```
Comment (chriseth) [inline test/libsolidity/util/TestFunctionCall.cpp:235]:
```suggestion
			stream << _linePrefix << " - " << reaction << ".";
```

File: test/libsolidity/SemanticTest.cpp (hunk 4 above)
Comment (aarlt) [reply test/libsolidity/SemanticTest.cpp:461]: I just added this to be able to test the "reactions". The name of this builtin is `smoke.reaction`, where the `SmokeHook` is setting a test-reaction for exactly that builtin.

File: test/libsolidity/SemanticTest.cpp (hunk 3 above)
Comment (aarlt) [reply test/libsolidity/SemanticTest.cpp:305]: I think that would be possible. So you prefer to have next to `Builtins` and `TestHooks` a third mechanism called e.g. `Effects`?

File: test/libsolidity/util/TestFileParser.cpp (hunk 5 above)
Comment (aarlt) [reply test/libsolidity/util/TestFileParser.cpp:279]: Thats true, I was not sure how this could be done nicely.

File: test/libsolidity/util/TestFileParser.cpp (hunk 1 above)
Comment (aarlt) [reply test/libsolidity/util/TestFileParser.cpp:702]: Newlines are definitely the most natural terminator. I think the newlines got already removed here, will check this.

File: test/libsolidity/util/TestFunctionCall.cpp (hunk 6 above)
Comment (aarlt) [reply test/libsolidity/util/TestFunctionCall.cpp:235]: Are you sure that the reactions should be outside the isoltest specific comments?

File: test/libsolidity/SemanticTest.cpp (hunk 3 above)
Comment (aarlt) [reply test/libsolidity/SemanticTest.cpp:305]: I mean, if I understood you correctly, you want to decouple this mechanism from the `TestHook`'s right?

File: test/libsolidity/SemanticTest.cpp (hunk 4 above)
Comment (chriseth) [reply test/libsolidity/SemanticTest.cpp:461]: I think it is fine to test them with actual events once we have that.

File: test/libsolidity/SemanticTest.cpp (hunk 3 above)
Comment (chriseth) [reply test/libsolidity/SemanticTest.cpp:305]: Yes

File: test/libsolidity/util/TestFunctionCall.cpp (hunk 6 above)
Comment (chriseth) [reply test/libsolidity/util/TestFunctionCall.cpp:235]: Ah sorry! Of course we need the slashes!

File: test/libsolidity/util/TestFileParser.cpp (hunk 5 above)
Comment (chriseth) [reply test/libsolidity/util/TestFileParser.cpp:279]: I don't understand - can't you use `-` (or it's token name) instead of `Token::Reaction`?

**Comment by aarlt** [comment]
Replaced by #11048.
